### DISCUSSION TITLE
server should support tools and tool_choice parameter

### DISCUSSION THREAD

**Discussion by segmond** [ggml-org/llama.cpp#6429] [Ideas]
When utilizing function calling with GPT, one can pass in the tools and tool_choice.   server doesn't support this and throws an error.    function calling is very nice and adds major usefulness to the API, should this be supported?

**Comment by Jeximo** [comment]
>  function calling is very nice and adds major usefulness to the API, should this be supported?

See [agent example (w/ Tools!) for server](https://github.com/ggerganov/llama.cpp/pull/6389#issue-2215943568), but keep in mind: it's a *work in progress*, and experimental.

**Comment by ngxson** [comment]
There's a WIP that I made here: https://github.com/ggerganov/llama.cpp/pull/5695

However, this PR is currently paused. We're currently focus on server stability, so this feature will not come very soon.
